affectedmoduledetector/src/main/kotlin/com/dropbox/affectedmoduledetector/util/File.kt

```diff
@@ -0,0 +1,24 @@
+package com.dropbox.affectedmoduledetector.util
+
+import java.io.File
+
+/**
+ * Converts a [String] representation of a relative [File] path to sections based on the OS
+ * specific separator character.
+ */
+fun String.toPathSections(rootProjectDir: File, gitRootDir: File): List<String> {
+    val realSections = toOsSpecificPath()
+        .split(File.separatorChar)
+        .toMutableList()
+    val projectRelativeDirectorySections = rootProjectDir
+        .toRelativeString(gitRootDir)
```

Comment:
Can you explain how the projects root directory and the git root directory differ? Seems like they should be the same

Reply:
> Can you explain how the projects root directory and the git root directory differ? Seems like they should be the same

Sorry I'm on mobile at the moment so can't back this up properly but.... If I remember correctly this was because the project I was working on had customised their Gradle project layout via `settings.gradle.kts`. 

We had a structure such that the root and sub projects weren't structured like a "normal" Gradle project would be but instead nested deeply in file system (99% of the projects I have worked on follow the standard Gradle structure but _we_ had to be different 😂). Gradle is cool with that since it uses `settings.gradle.kts` to resolve the `File` references to the root and sub projects but it breaks one of the assumptions that AMD uses.

Given that it was nonstandard but still valid Gradle I felt we should support those projects. Otherwise AMD reports false negatives.

So the git root directory would be, on the file system, the top most directory but the Gradle root project will be nested somewhere underneath it.

Sorry if that doesn't make sense I can draw a better diagram when I'm back?

Reply:
Yeah, diagram would be great.

I'm down for supporting weird setups but we have to kind of decide if its totally worth it for super weird setups ha. Otherwise this seems like a good use of the `baseDir` property that exists on our plugin (https://github.com/dropbox/AffectedModuleDetector/blob/6c2b902643507c510d7a9378a67a6e3411ce74af/affectedmoduledetector/src/main/kotlin/com/dropbox/affectedmoduledetector/AffectedModuleConfiguration.kt#L59) which seems to have a similar effect?

Reply:
Sorry for the delay here's a diagram to explain what's in my head.

A "traditional" Gradle project will look like this:

```
my-app <-- projectRootDir & gitRootDir
|
-- .git
|
-- module-1
|
-- module-2
|
-- module-3
|
-- settings.gradle.kts
```

The `.git` folder is a child of the `my-app` folder on the file system and thus the `my-app` folder is the `gitRootDir`. There are **no** customisations in `settings.gradle.kts` other than to `include` the modules `module-1`, `module-2`, and `module-3`. Gradle will also see `my-app` as the `projectRootDir`. This is what _most_ Android projects look like and is a fine and sensible layout to assume.

However I came across this setup at work:

```
android-project <-- gitRootDir
|
-- .git
|
-- diagrams
|
-- ci
|
-- proposals
|
-- my-app <-- projectRootDir
    |
    -- settings.gradle.kts
|
-- common
    |
    -- module-1
    |
    -- module-2
    |
    -- module-3
```

In this set up the `gitRootDir` and the `rootProjectDir` are not the same.

Having just typed all of that and also looking through the code again I think there is a simpler explanation. Whilst the above is still true.... there _can be_ exotic projects..... the more simpler explanation is that [I must have copied this over from the existing implementation](https://github.com/dropbox/AffectedModuleDetector/pull/172/files/b2acbc457bb3e4108014aea0d70f3100d518d79b#diff-8e77e2aac0662270a43112585ae885241facf304b722ba7c936b8409c4474db4L72). :laughing: 

Also..... regarding `baseDir`. That is nullable (and thus optional) in the AMD Configuration so we can't rely on it being set unless we go for an API change - I am taking the view that the Configuration forms part of the API for this :shrug: .

Reply:
Yeah, to me this feels like a totally separate issue than windows support so is it possible to pull this logic related to git out? I see what you are saying but I think its best to continue this discussion in its own PR rather than continue to block this one if you agree

On the point of nullability, if a user doesn't configure their project correctly then we might not want to worry about internally handling all the edge cases :)

Reply:
>Yeah, to me this feels like a totally separate issue than windows support so is it possible to pull this logic related to git out? I see what you are saying but I think its best to continue this discussion in its own PR rather than continue to block this one if you agree

Sorry I think I have sent us in the wrong direction here with talk about the alternative / exotic Gradle projects. I do totally agree that supporting these types of projects is out of scope of this PR. In fact in my original PR (#77) I included those additional changes but in this PR I have removed all but the Windows support since I agree the PRs should be focussed on one thing.

However with regards to `Files.kt` this is extracted from the code in `ProjectGraph::findContainingProject` and `ProjectGraph::findProjectRelativeDir` from the `main` branch. The reason I have done this is because I needed this functionality in two places: `ProjectGraph` and `AffectedModuleDetector`. Even in the `main` branch there is the use of `gitRootDir` and `rootProjectDir`. The modification I make is only to support Windows.

![image](https://user-images.githubusercontent.com/4974178/199354620-3ce5b715-ff15-4d50-8536-87d268b25701.png)

Again, massively sorry I gave the wrong impression in my first comment. This _isn't_ to support exotic project set ups. I was planning on raising a PR for that in the future but it's not this one. I will raise a separate issue about the exotic projects and we can continue discussions there. :bow:

Reply:
Alright, sounds good. Thanks for the clarification and the work on this!